experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,662 @@
+    l = length(v)
+    x = 1
+    for i in 1:l
+        x = x*factorial(v[i])
+    end
+    return Int(factorial(n)/x)
+end
+
+##########################
+#Invariant Rings of Reductive groups
+##########################
+mutable struct InvariantRing
+    field::Field
+    poly_ring::MPolyDecRing #graded
+    
+    group::ReductiveGroup
+    representation::RepresentationReductiveGroup
+    
+    fundamental::Vector{MPolyDecRingElem}
+    primary::Vector{MPolyDecRingElem}
+    secondary::Vector{MPolyDecRingElem}
+    
+    reynolds_operator::Function
+    
+    NullConeIdeal::MPolyIdeal
+    
+    #Invariant ring of reductive group G (in representation R), no other input.
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        if isdefined(R, :weights)
+            n = length(weights(R))
+            z.field = field(group(R))
+            super_ring, __ = graded_polynomial_ring(field(group(R)), "X"=>1:n)
+            z.poly_ring = super_ring
+            z.group = group(R)
+            z.representation = R
+            #z.fundamental = torus_invariants_fast(weights(R), ring)
+            return z
+        end
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = graded_polynomial_ring(G.field, "X" => 1:n)
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+    
+    #to compute invariant ring ring^G where G is the reductive group of R. 
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyDecRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        if isdefined(R, :weights)
+            #dosomething
+        end
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        z.poly_ring = ring
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+end
+
+invariant_ring(R::RepresentationReductiveGroup) = InvariantRing(R)
+invariant_ring(ring::MPolyDecRing, R::RepresentationReductiveGroup) = InvariantRing(R, ring)
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+poly_ring(R::InvariantRing) = R.poly_ring
+
+function fundamental_invariants(z::InvariantRing)
+    if isdefined(z, :fundamental)
+        return z.fundamental
+    else
+        R = z.representation
+        if group(group(R))[1] == :torus && isdefined(R, :weights)
+            z.fundamental = torus_invariants_fast(weights(R), poly_ring(z))
+            return z.fundamental
+        end
+        I, M = proj_of_image_ideal(R.group, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(R.group, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, R.group, z.poly_ring, M, I, z.reynolds_operator)
+        return z.fundamental
+    end
+end
```

Comment:
Second step for using attributes is here:
```suggestion
@attr Vector{MPolyDecRingElem} function fundamental_invariants(z::InvariantRing)
    R = z.representation
    if group(group(R))[1] == :torus && isdefined(R, :weights)
        return torus_invariants_fast(weights(R), poly_ring(z))
    end
    I, M = proj_of_image_ideal(R.group, R.rep_mat)
    z.NullConeIdeal = ideal(generators(R.group, I, R.rep_mat))
    return inv_generators(z.NullConeIdeal, R.group, z.poly_ring, M, I, z.reynolds_operator)
end
```

Reply:
Note that `Vector{MPolyDecRingElem}` indicates the return type which cannot be fully inferred for "attributes" like this one. However, it is suboptimal as `Vector{MPolyDecRingElem}` is not a concrete type. To make it concrete, consider adding "type parameters" to your `InvariantRing` type, just like the existing `struct InvRing` has. Please take a look at `src/InvariantTheory/types.jl:52` to see how that is done, and how it is used to make the members of `InvRing` have concrete types. Once that is in place, the signature of this attribute can be adapted to make it fully type stable.